ee/clickhouse/queries/cohort_query.py

```diff
@@ -0,0 +1,573 @@
+
+    def get_stopped_performing_event(self, prop: Property, prepend: str, idx: int) -> Tuple[str, Dict[str, Any]]:
+        event = (prop.event_type, prop.key)
+        before_column_name = f"stopped_event_condition_{prepend}_{idx}_before"
+        after_column_name = f"stopped_event_condition_{prepend}_{idx}_after"
+
+        entity_query, entity_params = self._get_entity(event, prepend, idx)
+        date_value = parse_and_validate_positive_integer(prop.time_value, "time_value")
+        date_param = f"{prepend}_date_{idx}"
+        date_interval = validate_interval(prop.time_interval)
+
+        seq_date_value = parse_and_validate_positive_integer(prop.seq_time_value, "time_value")
+        seq_date_param = f"{prepend}_seq_date_{idx}"
+        seq_date_interval = validate_interval(prop.seq_time_interval)
+
+        self._check_earliest_date([(date_value, date_interval), (seq_date_value, seq_date_interval)])
+
+        before_field = f"countIf(timestamp > now() - INTERVAL %({date_param})s {date_interval} - INTERVAL %({seq_date_param})s {seq_date_interval} AND timestamp < now() - INTERVAL %({date_param})s {date_interval} AND {entity_query}) AS {before_column_name}"
+        after_field = f"countIf(timestamp > now() - INTERVAL %({date_param})s {date_interval} AND timestamp < now() AND {entity_query}) AS {after_column_name}"
+
+        self._fields.append(before_field)
+        self._fields.append(after_field)
+
+        return (
+            f"({after_column_name} = 0 AND {before_column_name} > 0)",
+            {f"{date_param}": date_value, f"{seq_date_param}": seq_date_value, **entity_params},
+        )
+
+    def get_restarted_performing_event(self, prop: Property, prepend: str, idx: int) -> Tuple[str, Dict[str, Any]]:
+        event = (prop.event_type, prop.key)
+        before_column_name = f"restarted_event_condition_{prepend}_{idx}_before"
+        after_column_name = f"restarted_event_condition_{prepend}_{idx}_after"
+
+        entity_query, entity_params = self._get_entity(event, prepend, idx)
+        date_value = parse_and_validate_positive_integer(prop.time_value, "time_value")
+        date_param = f"{prepend}_date_{idx}"
+        date_interval = validate_interval(prop.time_interval)
+
+        seq_date_value = parse_and_validate_positive_integer(prop.seq_time_value, "time_value")
+        seq_date_param = f"{prepend}_seq_date_{idx}"
+        seq_date_interval = validate_interval(prop.seq_time_interval)
+
+        self._check_earliest_date([(date_value, date_interval), (seq_date_value, seq_date_interval)])
+
+        before_field = f"countIf(timestamp > now() - INTERVAL %({date_param})s {date_interval} - INTERVAL %({seq_date_param})s {seq_date_interval} AND timestamp < now() - INTERVAL %({date_param})s {date_interval} AND {entity_query}) AS {before_column_name}"
+        after_field = f"countIf(timestamp > now() - INTERVAL %({date_param})s {date_interval} AND timestamp < now() AND {entity_query}) AS {after_column_name}"
+
```

Comment:
The earliest date here should be an "all time" value, and we should check that the event occurred once before the `before field`. Otherwise, it isn't necessarily a "restarting" condition.